src/python/pants/core/target_types.py

```diff
@@ -206,5 +216,142 @@ class GenericTarget(Target):
+    `["resources:logo"]`.
+
+    This is useful to include any loose files, like data files, image assets, or config files.
+
+    This will ignore any targets that are not `files` or `relocated_files` targets. If you instead
+    want those files included in any packages specified in the `packages` field for this target,
+    then use a `resources` target and have the original package (e.g. the `python_library`)
+    depend on the resources.
+    """
+
+    alias = "files"
+
+
+class ArchiveFormatField(StringField):
+    """The type of archive file to be generated.
+
+    Note that `zip` files will not be compressed.
+    """
+
+    alias = "format"
+    valid_choices = ArchiveFormat
+    required = True
+    value: str
+
+
+class ArchiveTarget(Target):
+    """An archive (e.g. zip file) containing loose files and/or packages built via `./pants
+    package`."""
+
+    alias = "archive"
+    core_fields = (*COMMON_TARGET_FIELDS, ArchivePackages, ArchiveFiles, ArchiveFormatField)
+
+
+@dataclass(frozen=True)
+class ArchiveFieldSet(PackageFieldSet):
+    required_fields = (ArchiveFormatField,)
+
+    packages: ArchivePackages
+    files: ArchiveFiles
+    format_field: ArchiveFormatField
+
+
+@rule(level=LogLevel.DEBUG)
+async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+    package_targets = await MultiGet(
+        Get(
+            WrappedTarget,
+            AddressInput,
+            AddressInput.parse(v, relative_to=field_set.address.spec_path),
+        )
+        for v in field_set.packages.value or ()
+    )
+    package_field_sets_per_target = await Get(
+        FieldSetsPerTarget,
+        FieldSetsPerTargetRequest(
+            PackageFieldSet,
+            (wrapped_tgt.target for wrapped_tgt in package_targets),
+        ),
+    )
+    packages = await MultiGet(
+        Get(BuiltPackage, PackageFieldSet, field_set)
+        for field_set in package_field_sets_per_target.field_sets
+    )
+
+    files_targets = await MultiGet(
+        Get(
+            WrappedTarget,
+            AddressInput,
+            AddressInput.parse(v, relative_to=field_set.address.spec_path),
+        )
+        for v in field_set.files.value or ()
+    )
+    files_sources = await MultiGet(
+        Get(
+            HydratedSources,
+            HydrateSourcesRequest(
+                wrapped_tgt.target.get(Sources),
+                for_sources_types=(FilesSources,),
+                enable_codegen=True,
+            ),
+        )
+        for wrapped_tgt in files_targets
+    )
+
+    input_snapshot = await Get(
+        Snapshot,
+        MergeDigests(
+            (
+                *(package.digest for package in packages),
+                *(sources.snapshot.digest for sources in files_sources),
+            )
+        ),
+    )
+    file_ending = field_set.format_field.value
+    output_filename = os.path.join(
+        field_set.address.spec_path.replace(os.sep, "."),
```

Comment:
It's possible that if we're going to take this approach (rather than having users choose explicit names for archives) that we should switch to recreating the nested directory structure that they were located in rather than doing the `replace(os.sep, ".")` dance.

Filenames have length limits (and paths too, but the names are more constraining in practice), and particularly in docker or on encrypted filesystems they can be quite low: on the order of 135 chars. Preserving the directory structure makes it much harder to hit those limits.

Reply:
Benjy originally did the full path, and it ended up being clunky. You would have to traverse multiple empty directories to get the asset you actually want.

I think this is a decent default, even though it has some risks. If a user hits too long of a filename, they can use the override on each binary target, which I think you were proposing above.

Reply:
I expect that even with the override, the override would be a relative path, rather than the absolute output path?

Reply:
I think we would want to allow it to be relative to the build root. Otherwise, you have no mechanism to get `src.python.pants/pants.pex` to simply be `pants.pex` in your archive. Or to be `packages/pants.pex`. It's not adequate to only rename the `base_name`.

This gets a little bit into `relocated_files` area of whether the relocation should live on the original `python_binary`, at consumers, or some intermediate thing. Because often the consumer _is_ the user via `./pants package`, I think we'd want the remapping logic to live on the original package target.

Reply:
There is some risk for collisions, of course. But Everyone Is An Adult here imo. Our default is sensible and guarantees no collisions, and we're also giving this advanced override if you are okay with the risk.

Reply:
Couldn't you depend on the archive with `relocated_files` to relocate it?

Reply:
> And the output filename in dist isn't the name you need to ship it with.

Yes, but if you have an `archive` that contains another `archive` that itself contains a `.pex`, this gets tricky to need to relocate those inner packages. You would need to extract the archive built by Pants, reshuffle all the packages, and then re-create the archive.

The other risk is that the archive names are less stable than a user may want. They may not want it so that renaming a folder or file will result in the package's name changing. They might prefer using a stable, hardcoded path.

I strongly suspect we need to allow setting the output path on any package. But again, our default is sensible and guaranteed to be unambiguous.

Example:

```python
archive(
  packages=[":pkg1"],
  output="zip",
  output_path="zipped_projects/foo.zip",  # This gets prefixed by `--distdir` still.
)
```

Reply:
> Yes, but if you have an `archive` that contains another `archive` that itself contains a `.pex`, this gets tricky to need to relocate those inner packages. You would need to extract the archive built by Pants, reshuffle all the packages, and then re-create the archive.

_That_ should work via `relocated_files` though, right? `archive <- relocated_files <- archive`?

Reply:
No. Relocated_files do not work on package types. If you put `relocated_files(original_files=[":pkg"])`, it will _not_ build the package for you. That field only works with `files` targets.